diff -Naur hdhomerun_config_gui.orig/Makefile.am hdhomerun_config_gui/Makefile.am
--- hdhomerun_config_gui.orig/Makefile.am	2014-03-13 14:46:35.747681686 +0100
+++ hdhomerun_config_gui/Makefile.am	2014-03-13 14:46:52.360772999 +0100
@@ -1,6 +1,6 @@
 ## Process this file with automake to produce Makefile.in
 
-SUBDIRS = ../libhdhomerun src po
+SUBDIRS = src po
 
 EXTRA_DIST = \
 	autogen.sh \
diff -Naur hdhomerun_config_gui.orig/src/AppInclude.h hdhomerun_config_gui/src/AppInclude.h
--- hdhomerun_config_gui.orig/src/AppInclude.h	2014-03-13 14:46:35.747681686 +0100
+++ hdhomerun_config_gui/src/AppInclude.h	2014-03-13 14:51:52.511417891 +0100
@@ -15,7 +15,7 @@
 #  include <CoreFoundation/CoreFoundation.h>
 #endif
 
-#include "hdhomerun.h"
+#include <hdhomerun/hdhomerun.h>
 
 #include <gtk/gtk.h>
 #ifdef __WINDOWS__
diff -Naur hdhomerun_config_gui.orig/src/Makefile.am hdhomerun_config_gui/src/Makefile.am
--- hdhomerun_config_gui.orig/src/Makefile.am	2014-03-13 14:46:35.747681686 +0100
+++ hdhomerun_config_gui/src/Makefile.am	2014-03-13 14:50:48.000065015 +0100
@@ -1,7 +1,6 @@
 ## Process this file with automake to produce Makefile.in
 
 INCLUDES = \
-	-I../../libhdhomerun \
 	-DPACKAGE_DATA_DIR=\""$(datadir)"\" \
 	-DPACKAGE_LOCALE_DIR=\""$(prefix)/$(DATADIRNAME)/locale"\" \
 	@PACKAGE_CFLAGS@
@@ -18,16 +17,7 @@
 	Viewer.cpp Viewer.h \
 	HDHRConfig.cpp HDHRConfig.h
 
-install-exec-local:
-	test -z "$(libdir)" || $(MKDIR_P) "$(DESTDIR)$(libdir)"
-	$(INSTALL) ../../libhdhomerun/libhdhomerun.so $(DESTDIR)$(libdir)
-	$(INSTALL) ../../libhdhomerun/hdhomerun_config $(DESTDIR)$(bindir)
-
-uninstall-local:
-	-rm -f "$(DESTDIR)$(libdir)/libhdhomerun.so"
-	-rm -f "$(DESTDIR)$(bindir)/hdhomerun_config" 
-	
 hdhomerun_config_gui_LDADD = \
-	-L../../libhdhomerun -lhdhomerun \
+	-lhdhomerun \
 	@PACKAGE_LIBS@ $(INTLLIBS)
 

