diff -Naur gpointing-device-settings-1.5.1.orig/configure.ac gpointing-device-settings-1.5.1/configure.ac
--- gpointing-device-settings-1.5.1.orig/configure.ac	2014-02-12 13:57:49.256163071 +0100
+++ gpointing-device-settings-1.5.1/configure.ac	2014-02-12 13:58:05.404276945 +0100
@@ -4,7 +4,6 @@
 AC_INIT(gpointing-device-settings, 1.5.1,
         [http://bugzilla.gnome.org/enter_bug.cgi?product=GPointingDeviceSettings])
 AM_INIT_AUTOMAKE([1.9 tar-pax])
-AM_INIT_AUTOMAKE([foreign])
 m4_ifdef([AM_SILENT_RULES],[AM_SILENT_RULES([yes])])
 AM_CONFIG_HEADER(config.h)
 AC_CONFIG_MACRO_DIR([m4])
