diff -crB clutk/clutk/ctk-perspective-correct-rendering.c clutk2/clutk/ctk-perspective-correct-rendering.c
*** clutk/clutk/ctk-perspective-correct-rendering.c	Fri Oct  1 18:56:45 2010
--- clutk2/clutk/ctk-perspective-correct-rendering.c	Fri Oct  1 18:55:53 2010
***************
*** 57,66 ****
  {
    gfloat VtxBuffer[] = {
       // X    Y       Z        W       S              T              R              Q              R                      G                        B                       A
!      V[0].x, V[0].y, V[0].z, 1.0f,  V[0].s/V[0].w, V[0].t/V[0].w, V[0].r/V[0].w, V[0].q/V[0].w, V[0].color.red/255.0f, V[0].color.green/255.0f, V[0].color.blue/255.0f, V[0].color.alpha/255.0f,
!      V[1].x, V[1].y, V[1].z, 1.0f,  V[1].s/V[1].w, V[1].t/V[1].w, V[1].r/V[1].w, V[1].q/V[1].w, V[1].color.red/255.0f, V[1].color.green/255.0f, V[1].color.blue/255.0f, V[1].color.alpha/255.0f,
!      V[2].x, V[2].y, V[2].z, 1.0f,  V[2].s/V[2].w, V[2].t/V[2].w, V[2].r/V[2].w, V[2].q/V[2].w, V[2].color.red/255.0f, V[2].color.green/255.0f, V[2].color.blue/255.0f, V[2].color.alpha/255.0f,
!      V[3].x, V[3].y, V[3].z, 1.0f,  V[3].s/V[3].w, V[3].t/V[3].w, V[3].r/V[3].w, V[3].q/V[3].w, V[3].color.red/255.0f, V[3].color.green/255.0f, V[3].color.blue/255.0f, V[3].color.alpha/255.0f,
    };
  
  //   printf("V0: %f, %f, %f, %f \n", V[0].x, V[0].y, V[0].z, V[0].w);
--- 57,66 ----
  {
    gfloat VtxBuffer[] = {
       // X    Y       Z        W       S              T              R              Q              R                      G                        B                       A
!      V[0].x, V[0].y, V[0].z, 1.0f,  V[0].s/V[0].w, V[0].t/V[0].w, V[0].r/V[0].w, V[0].q/V[0].w, cogl_color_get_red(&V[0].color), cogl_color_get_green(&V[0].color), cogl_color_get_blue(&V[0].color), cogl_color_get_alpha(&V[0].color),
!      V[1].x, V[1].y, V[1].z, 1.0f,  V[1].s/V[1].w, V[1].t/V[1].w, V[1].r/V[1].w, V[1].q/V[1].w, cogl_color_get_red(&V[1].color), cogl_color_get_green(&V[1].color), cogl_color_get_blue(&V[1].color), cogl_color_get_alpha(&V[1].color),
!      V[2].x, V[2].y, V[2].z, 1.0f,  V[2].s/V[2].w, V[2].t/V[2].w, V[2].r/V[2].w, V[2].q/V[2].w, cogl_color_get_red(&V[2].color), cogl_color_get_green(&V[2].color), cogl_color_get_blue(&V[2].color), cogl_color_get_alpha(&V[2].color),
!      V[3].x, V[3].y, V[3].z, 1.0f,  V[3].s/V[3].w, V[3].t/V[3].w, V[3].r/V[3].w, V[3].q/V[3].w, cogl_color_get_red(&V[3].color), cogl_color_get_green(&V[3].color), cogl_color_get_blue(&V[3].color), cogl_color_get_alpha(&V[3].color),
    };
  
  //   printf("V0: %f, %f, %f, %f \n", V[0].x, V[0].y, V[0].z, V[0].w);
***************
*** 133,151 ****
  
    //CHECKGL( glDrawArrays(GL_QUADS, 0, 4) );
    glBegin(GL_QUADS);
!   glColor4f(V[0].color.red/255.0f, V[0].color.green/255.0f, V[0].color.blue/255.0f, V[0].color.alpha/255.0f);
    glMultiTexCoord4f(GL_TEXTURE0, V[0].s/V[0].w, V[0].t/V[0].w, V[0].r/V[0].w, V[0].q/V[0].w);
    glVertex4f(V[0].x, V[0].y, V[0].z, 1.0f);
    
!   glColor4f(V[1].color.red/255.0f, V[1].color.green/255.0f, V[1].color.blue/255.0f, V[1].color.alpha/255.0f);
    glMultiTexCoord4f(GL_TEXTURE0, V[1].s/V[1].w, V[1].t/V[1].w, V[1].r/V[1].w, V[1].q/V[1].w);
    glVertex4f(V[1].x, V[1].y, V[1].z, 1.0f);
    
!   glColor4f(V[2].color.red/255.0f, V[2].color.green/255.0f, V[2].color.blue/255.0f, V[2].color.alpha/255.0f);
    glMultiTexCoord4f(GL_TEXTURE0, V[2].s/V[2].w, V[2].t/V[2].w, V[2].r/V[2].w, V[2].q/V[2].w);
    glVertex4f(V[2].x, V[2].y, V[2].z, 1.0f);
    
!   glColor4f(V[3].color.red/255.0f, V[3].color.green/255.0f, V[3].color.blue/255.0f, V[3].color.alpha/255.0f);
    glMultiTexCoord4f(GL_TEXTURE0, V[3].s/V[3].w, V[3].t/V[3].w, V[3].r/V[3].w, V[3].q/V[3].w);
    glVertex4f(V[3].x, V[3].y, V[3].z, 1.0f);
    glEnd();  
--- 133,151 ----
  
    //CHECKGL( glDrawArrays(GL_QUADS, 0, 4) );
    glBegin(GL_QUADS);
!   glColor4f(cogl_color_get_red(&V[0].color), cogl_color_get_green(&V[0].color), cogl_color_get_blue(&V[0].color), cogl_color_get_alpha(&V[0].color));
    glMultiTexCoord4f(GL_TEXTURE0, V[0].s/V[0].w, V[0].t/V[0].w, V[0].r/V[0].w, V[0].q/V[0].w);
    glVertex4f(V[0].x, V[0].y, V[0].z, 1.0f);
    
!   glColor4f(cogl_color_get_red(&V[1].color), cogl_color_get_green(&V[1].color), cogl_color_get_blue(&V[1].color), cogl_color_get_alpha(&V[1].color));
    glMultiTexCoord4f(GL_TEXTURE0, V[1].s/V[1].w, V[1].t/V[1].w, V[1].r/V[1].w, V[1].q/V[1].w);
    glVertex4f(V[1].x, V[1].y, V[1].z, 1.0f);
    
!   glColor4f(cogl_color_get_red(&V[2].color), cogl_color_get_green(&V[2].color), cogl_color_get_blue(&V[2].color), cogl_color_get_alpha(&V[2].color));
    glMultiTexCoord4f(GL_TEXTURE0, V[2].s/V[2].w, V[2].t/V[2].w, V[2].r/V[2].w, V[2].q/V[2].w);
    glVertex4f(V[2].x, V[2].y, V[2].z, 1.0f);
    
!   glColor4f(cogl_color_get_red(&V[3].color), cogl_color_get_green(&V[3].color), cogl_color_get_blue(&V[3].color), cogl_color_get_alpha(&V[3].color));
    glMultiTexCoord4f(GL_TEXTURE0, V[3].s/V[3].w, V[3].t/V[3].w, V[3].r/V[3].w, V[3].q/V[3].w);
    glVertex4f(V[3].x, V[3].y, V[3].z, 1.0f);
    glEnd();  
***************
*** 184,193 ****
  {
    gfloat VtxBuffer[] = {
       // X    Y       Z        W       S              T              R              Q              R                      G                        B                       A
!      V[0].x, V[0].y, V[0].z, 1.0f,  V[0].s/V[0].w, V[0].t/V[0].w, V[0].r/V[0].w, V[0].q/V[0].w, V[0].color.red/255.0f, V[0].color.green/255.0f, V[0].color.blue/255.0f, V[0].color.alpha/255.0f,
!      V[1].x, V[1].y, V[1].z, 1.0f,  V[1].s/V[1].w, V[1].t/V[1].w, V[1].r/V[1].w, V[1].q/V[1].w, V[1].color.red/255.0f, V[1].color.green/255.0f, V[1].color.blue/255.0f, V[1].color.alpha/255.0f,
!      V[2].x, V[2].y, V[2].z, 1.0f,  V[2].s/V[2].w, V[2].t/V[2].w, V[2].r/V[2].w, V[2].q/V[2].w, V[2].color.red/255.0f, V[2].color.green/255.0f, V[2].color.blue/255.0f, V[2].color.alpha/255.0f,
!      V[3].x, V[3].y, V[3].z, 1.0f,  V[3].s/V[3].w, V[3].t/V[3].w, V[3].r/V[3].w, V[3].q/V[3].w, V[3].color.red/255.0f, V[3].color.green/255.0f, V[3].color.blue/255.0f, V[3].color.alpha/255.0f,
    };
  
  //   printf("V0: %f, %f, %f, %f \n", V[0].x, V[0].y, V[0].z, V[0].w);
--- 184,193 ----
  {
    gfloat VtxBuffer[] = {
       // X    Y       Z        W       S              T              R              Q              R                      G                        B                       A
!      V[0].x, V[0].y, V[0].z, 1.0f,  V[0].s/V[0].w, V[0].t/V[0].w, V[0].r/V[0].w, V[0].q/V[0].w, cogl_color_get_red(&V[0].color), cogl_color_get_green(&V[0].color), cogl_color_get_blue(&V[0].color), cogl_color_get_alpha(&V[0].color),
!      V[1].x, V[1].y, V[1].z, 1.0f,  V[1].s/V[1].w, V[1].t/V[1].w, V[1].r/V[1].w, V[1].q/V[1].w, cogl_color_get_red(&V[1].color), cogl_color_get_green(&V[1].color), cogl_color_get_blue(&V[1].color), cogl_color_get_alpha(&V[1].color),
!      V[2].x, V[2].y, V[2].z, 1.0f,  V[2].s/V[2].w, V[2].t/V[2].w, V[2].r/V[2].w, V[2].q/V[2].w, cogl_color_get_red(&V[2].color), cogl_color_get_green(&V[2].color), cogl_color_get_blue(&V[2].color), cogl_color_get_alpha(&V[2].color),
!      V[3].x, V[3].y, V[3].z, 1.0f,  V[3].s/V[3].w, V[3].t/V[3].w, V[3].r/V[3].w, V[3].q/V[3].w, cogl_color_get_red(&V[3].color), cogl_color_get_green(&V[3].color), cogl_color_get_blue(&V[3].color), cogl_color_get_alpha(&V[3].color),
    };
  
  //   printf("V0: %f, %f, %f, %f \n", V[0].x, V[0].y, V[0].z, V[0].w);
***************
*** 264,282 ****
  
    //CHECKGL( glDrawArrays(GL_QUADS, 0, 4) );
    glBegin(GL_QUADS);
!   glColor4f(V[0].color.red/255.0f, V[0].color.green/255.0f, V[0].color.blue/255.0f, V[0].color.alpha/255.0f);
    glMultiTexCoord4f(GL_TEXTURE0, V[0].s/V[0].w, V[0].t/V[0].w, V[0].r/V[0].w, V[0].q/V[0].w);
    glVertex4f(V[0].x, V[0].y, V[0].z, 1.0f);
    
!   glColor4f(V[1].color.red/255.0f, V[1].color.green/255.0f, V[1].color.blue/255.0f, V[1].color.alpha/255.0f);
    glMultiTexCoord4f(GL_TEXTURE0, V[1].s/V[1].w, V[1].t/V[1].w, V[1].r/V[1].w, V[1].q/V[1].w);
    glVertex4f(V[1].x, V[1].y, V[1].z, 1.0f);
    
!   glColor4f(V[2].color.red/255.0f, V[2].color.green/255.0f, V[2].color.blue/255.0f, V[2].color.alpha/255.0f);
    glMultiTexCoord4f(GL_TEXTURE0, V[2].s/V[2].w, V[2].t/V[2].w, V[2].r/V[2].w, V[2].q/V[2].w);
    glVertex4f(V[2].x, V[2].y, V[2].z, 1.0f);
    
!   glColor4f(V[3].color.red/255.0f, V[3].color.green/255.0f, V[3].color.blue/255.0f, V[3].color.alpha/255.0f);
    glMultiTexCoord4f(GL_TEXTURE0, V[3].s/V[3].w, V[3].t/V[3].w, V[3].r/V[3].w, V[3].q/V[3].w);
    glVertex4f(V[3].x, V[3].y, V[3].z, 1.0f);
    glEnd();
--- 264,282 ----
  
    //CHECKGL( glDrawArrays(GL_QUADS, 0, 4) );
    glBegin(GL_QUADS);
!   glColor4f(cogl_color_get_red(&V[0].color), cogl_color_get_green(&V[0].color), cogl_color_get_blue(&V[0].color), cogl_color_get_alpha(&V[0].color));
    glMultiTexCoord4f(GL_TEXTURE0, V[0].s/V[0].w, V[0].t/V[0].w, V[0].r/V[0].w, V[0].q/V[0].w);
    glVertex4f(V[0].x, V[0].y, V[0].z, 1.0f);
    
!   glColor4f(cogl_color_get_red(&V[1].color), cogl_color_get_green(&V[1].color), cogl_color_get_blue(&V[1].color), cogl_color_get_alpha(&V[1].color));
    glMultiTexCoord4f(GL_TEXTURE0, V[1].s/V[1].w, V[1].t/V[1].w, V[1].r/V[1].w, V[1].q/V[1].w);
    glVertex4f(V[1].x, V[1].y, V[1].z, 1.0f);
    
!   glColor4f(cogl_color_get_red(&V[2].color), cogl_color_get_green(&V[2].color), cogl_color_get_blue(&V[2].color), cogl_color_get_alpha(&V[2].color));
    glMultiTexCoord4f(GL_TEXTURE0, V[2].s/V[2].w, V[2].t/V[2].w, V[2].r/V[2].w, V[2].q/V[2].w);
    glVertex4f(V[2].x, V[2].y, V[2].z, 1.0f);
    
!   glColor4f(cogl_color_get_red(&V[3].color), cogl_color_get_green(&V[3].color), cogl_color_get_blue(&V[3].color), cogl_color_get_alpha(&V[3].color));
    glMultiTexCoord4f(GL_TEXTURE0, V[3].s/V[3].w, V[3].t/V[3].w, V[3].r/V[3].w, V[3].q/V[3].w);
    glVertex4f(V[3].x, V[3].y, V[3].z, 1.0f);
    glEnd();
